### PR TITLE
Add `scipy.special.zetac` to cupyx

### PR DESCRIPTION
**Pull request by Raghav323**
This PR aims to add `scipy.special.zetac` to cupy . 
Linked Issue : #6324

### CODE REVIEW

File: tests/cupyx_tests/scipy_tests/special_tests/test_zetac.py [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+        return scp.special.zetac(a)
+
+    @testing.for_all_dtypes(no_complex=True, no_bool=True)
+    @testing.numpy_cupy_allclose(atol=1e-5, rtol=1e-6, scipy_name='scp')
+    def test_linspace(self, xp, scp, dtype):
+        import scipy.special  # NOQA
+
+        a = numpy.linspace(-30, 30, 1000, dtype=dtype)
+
+        a = xp.asarray(a)
```
Comment (takagi) [inline tests/cupyx_tests/scipy_tests/special_tests/test_zetac.py:28]:
```suggestion
        a = xp.linspace(-30, 30, 1000, dtype=dtype)
```

File: tests/cupyx_tests/scipy_tests/special_tests/test_zetac.py [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+        return scp.special.zetac(a)
+
+    @testing.for_all_dtypes(no_complex=True, no_bool=True)
+    @testing.numpy_cupy_allclose(atol=1e-5, rtol=1e-6, scipy_name='scp')
+    def test_linspace_smallnegativex(self, xp, scp, dtype):
+        import scipy.special  # NOQA
+
+        a = numpy.linspace(-1, 0, 1000, dtype=dtype)
+
+        a = xp.asarray(a)
```
Comment (takagi) [inline tests/cupyx_tests/scipy_tests/special_tests/test_zetac.py:39]:
```suggestion
        a = xp.linspace(-1, 0, 1000, dtype=dtype)
```

File: tests/cupyx_tests/scipy_tests/special_tests/test_zetac.py [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+
+    @testing.for_all_dtypes(no_complex=True)
+    @testing.numpy_cupy_allclose(atol=1e-2, rtol=1e-3, scipy_name='scp')
+    def test_inf_and_nan(self, xp, scp, dtype):
+        import scipy.special  # NOQA
+
+        x = numpy.array([-numpy.inf, numpy.nan, numpy.inf]).astype(dtype)
+        a = numpy.tile(x, (3, 3))
+
+        a = xp.asarray(a)
```
Comment (takagi) [inline tests/cupyx_tests/scipy_tests/special_tests/test_zetac.py:58]: As other my comment, we can use `xp.array` and `xp.tile` here to have a code that works on both NumPy and CuPy.

**Comment by takagi** [comment]
The kernel function has branching by the value of `x`, which occurs divergence on GPU unlike that on CPU. If it is for numerical stability, the branching is fine; however, if it is intended to make early returns for efficiency, it wouldn't work as expected as the threads run through both the then-branch and else-branch and take the sum of these times.

File: tests/cupyx_tests/scipy_tests/special_tests/test_zetac.py [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+
+
+@testing.gpu
+@testing.with_requires('scipy')
+class TestZetac(unittest.TestCase):
+
+    @testing.for_all_dtypes(no_complex=True)
+    @testing.numpy_cupy_allclose(atol=1e-5, scipy_name='scp')
+    def test_arange(self, xp, scp, dtype):
+        import scipy.special  # NOQA
```
Comment (takagi) [inline tests/cupyx_tests/scipy_tests/special_tests/test_zetac.py:15]: We'd like to import modules on the top of files whenever it's possible. Would you make it as the following?
https://github.com/cupy/cupy/blob/c8bcae476916985074f5dca4a9f6a5d10458fccd/tests/cupyx_tests/scipy_tests/signal_tests/test_splines.py#L10-L13

**Comment by takagi** [comment]
Would you make a performance comparison against SciPy as #7474? You may use and modify the benchmark script there, which is the way we commonly take.

**Comment by Raghav323** [comment]
> For the record: The kernel function has branching by the value of `x`, which occurs divergence on GPU unlike that on CPU. If it is for numerical stability, the branching is fine; however, if it is intended to make early returns for efficiency, it wouldn't work as expected as the threads run through both the then-branch and else-branch and take the sum of these times.

Thanks! Noted :+1:

**Comment by Raghav323** [comment]
> Would you make a performance comparison against SciPy as #7474? You may use and modify the benchmark script there, which is the way we commonly take.


Thanks , will keep that in mind . Here are the benchmarks : 

Size | dtype | SciPy | CuPy
-----|-------|-------|-----
1000 | float16 | 0.021 ms | 0.038 ms
1000 | float32 | 0.044 ms | 0.031 ms
1000 | float64 | 0.028 ms | 0.036 ms
100000 | float16 | 2.712 ms | 0.074 ms
100000 | float32 | 2.073 ms | 0.077 ms
100000 | float64 | 1.815 ms | 0.090 ms
1000000 | float16 | 26.502 ms | 0.497 ms
1000000 | float32 | 22.906 ms | 0.492 ms
1000000 | float64 | 15.927 ms | 0.540 ms


<details>
<summary>Script</summary>

```
import cupy
import cupyx
import scipy
import numpy 
from cupyx.profiler import benchmark


n_warmup = 1
n_repeat = 10
dtype_set = ('float16','float32', 'float64')
n_set = (1000, 100000, 1000000)


def get_time(pref):
    cpu_time = pref.cpu_times.mean()
    gpu_time = pref.gpu_times.mean()
    return max(cpu_time, gpu_time) * 1000  # ms


def time(a):
    cp_a = cupy.array(a)
    times = []
    perf = benchmark(scipy.special.zetac, (a,),
                     n_warmup=n_warmup, n_repeat=n_repeat)
    times.append(get_time(perf))
    perf = benchmark(cupyx.scipy.special.zetac, (cp_a,),
                     n_warmup=n_warmup, n_repeat=n_repeat)
    times.append(get_time(perf))
    return times


print('size | dtype | SciPy | CuPy')
print('-----|-------|-------|-----')
for n in n_set:
    for dtype in dtype_set:
        a = numpy.random.random(size=n).astype(dtype) * 0.01
        times = time(a)
        print('{} | {} | {:.3f} ms | {:.3f} ms'.format(n, dtype, times[0], times[1]))

```

</details>

**Review by takagi — DISMISSED** [review]
LGTM!

File: tests/cupyx_tests/scipy_tests/special_tests/test_zetac.py [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+import cupyx.scipy.special  # NOQA
+import pytest
+
+try:
+    import scipy.special  # NOQA
+except ImportError:
+    pass
+
+
+@testing.gpu
```
Comment (takagi) [inline tests/cupyx_tests/scipy_tests/special_tests/test_zetac.py:13]: This decorator was here due to a historical reason and was removed in #7453.
```suggestion
```

**Comment by Raghav323** [comment]
Any hints on the CI failure ?

**Comment by Raghav323** [comment]
Hi , I was wondering if this PR could be merged now or there are still some fixes required from my side ? I believe the previous failure was because of me not taking in account the removal of gpu decorator in #7453 .

**Comment by Raghav323** [comment]
Also i noticed a small error from my side in one of the tests which makes it skip the test every time , Will fix it asap .

**Review by takagi — APPROVED** [review]
LGTM!

**Comment by takagi** [comment]
Thanks, @Raghav323!

**Pull request merged**